src/replication/replication_node_manager.cpp

```diff
@@ -61,44 +75,75 @@ namespace cubreplication
 
     void finalize ()
     {
+      wait_ha_tasks ();
+
       g_hostname.clear ();
 
       delete g_slave_node;
       g_slave_node = NULL;
       delete g_master_node;
       g_master_node = NULL;
 
-      // stream and stream file are interdependent, therefore first stop the stream
+      // we need to first stop the stream before destroying stream_file
       g_stream->stop ();
       delete g_stream_file;
       g_stream_file = NULL;
       delete g_stream;
       g_stream = NULL;
     }
 
-    void commute_to_master_state ()
+    void commute_to_master_state (const std::function<void (void)> &ha_transitions)
```

Comment:
there are two ha_transitions:
```c
  logtb_enable_update
  ha_Server_state = HA_SERVER_STATE_ACTIVE
```
```c
  logtb_enable_update
  state = css_transit_ha_server_state
  assert (state  == HA_SERVER_STATE_ACTIVE);
```

I think commute would be more clear without the ha_transitions lambda. It should need boolean force argument. Task execute should call logtb_enable_update + another css ha state function like this:

```c
void
css_commute_ha_state_or_another_proper_name (THREAD_ENTRY * thread_p, state desired_state, bool force)
{
  new_state = css_transit_ha_server_state (thread_p, desired_state);
  assert (new_state == desired_state);
  if (force)
    {
      ha_Server_state = desired_state;
    }
}
```

Also, please avoid big lambda functions directly in argument list. Better to do two statements:
```c
  auto a_lambda_name = [&ha_transitions] (cubthread::entry &context)
    {
      ...
    };
  cubthread::entry_task *promote_task = new cubthread::entry_callable_task (a_lambda_name);
```